assert: handle enum. symbol keys in deepStrictEqual

BridgeAR opened the pull request:
Adding support for enumerable symbol keys is consequent as we do compare symbols as primitives in deepEqual but we do not check them as keys.

I think this is important as I think we should try to prevent using enumerable symbols as private keys in our code and instead only use non-enumerable ones.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
assert

BridgeAR commented:
CITGM https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/964/

Trott commented on doc/api/assert.md:
```diff
@@ -43,12 +43,11 @@ Primitive values are compared with the [Abstract Equality Comparison][]
-throw an `AssertionError` because the properties on the [`RegExp`][] object are
-not enumerable:
+[`[[Prototype]]`][prototype-spec] of objects or enumerable own [`Symbol`][]
+properties — for such checks, consider using [`assert.deepStrictEqual()`][]
```
Micro-nit: Get rid of the `—` and make these two simple sentences instead:

> Only [enumerable "own" properties][] are considered. The [`assert.deepEqual()`][] implementation does not test the [`[[Prototype]]`][prototype-spec] of objects or enumerable own [`Symbol`][] properties. For such checks, consider using [`assert.deepStrictEqual()`][] instead.

Trott commented on doc/api/assert.md:
```diff
@@ -43,12 +43,11 @@ Primitive values are compared with the [Abstract Equality Comparison][]
-not enumerable:
+[`[[Prototype]]`][prototype-spec] of objects or enumerable own [`Symbol`][]
+properties — for such checks, consider using [`assert.deepStrictEqual()`][]
+instead. This can lead to some potentially surprising results. For example, the
```
Nit: New paragraph starting with `This can lead…`. Maybe also replace `This`. And get rid of `For example` as the next few words are `the following example`. And linkify `AssertionError` maybe? All told, maybe this?:

> `assert.deepEqual()` can have potentially surprising results. The following example does not throw an [`AssertioneError`][] because the properties on the [`RegExp`][] object are not enumerable:

Trott commented on doc/api/assert.md:
```diff
@@ -105,6 +104,9 @@ parameter is undefined, a default error message is assigned.
 changes:
   - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/XXXXX
+    description: Enumerable symbols are now compared as well.
```
Nano-nit: Remove `as well`?

BridgeAR commented:
Comments addressed and rebased due to conflicts.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/9974/

BridgeAR commented:
@jasnell It seems like something broke CITGM, would you mind taking a look?

jasnell commented:
broke in what way? I'm on an extremely slow in flight wifi for the next 9 hours so it may be better to have someone else take a look. @nodejs/citgm

BridgeAR commented:
@jasnell check the test names of the failures in e.g. https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/964/

refack commented:
RE: CITGM - npm was broken between https://github.com/nodejs/node/pull/15053 and https://github.com/nodejs/node/pull/15131
Rerun: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/966/

BridgeAR commented:
@refack urgs, you are right. Your CITGM fails as well though because you did not rebase to master^^

Here is a build rebased on master https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/967/

jasnell commented:
@BridgeAR ... can you rebase this?

BridgeAR commented:
@jasnell I will look into this soon. I missed a fast path and I want to think about how to check that in a nice way. That is the reason I did not yet land it^^.

BridgeAR commented:
I rebased this and I refactored the code to make sure the fast paths are also checking for the symbols. This required some more changes than before.

So PTAL

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10139/

BridgeAR commented:
Landed in 027bdae93250f0aa9047189946daa1824c223c96
